csrc/multidevice/communication.cpp

```diff
@@ -107,214 +201,132 @@ std::string Communication::toString(int indent) const {
   return ss.str();
 }
 
-Broadcast::Broadcast(CommParams params) : Communication(params, "broadcast") {}
-
-c10::intrusive_ptr<c10d::Work> Broadcast::post(
-    Communicator& comm,
-    std::optional<CommunicatorBackend> backend) {
-  post_common(*this, comm);
-
-  if (comm.deviceId() == params_.root) {
-    assertBufferCount(params_.src_bufs, 1);
-    if (params_.dst_bufs.size() == 1) {
-      doLocalCopy(params_.dst_bufs.at(0), params_.src_bufs.at(0));
-    } else {
-      assertBufferCount(params_.dst_bufs, 0);
-    }
-  } else {
-    assertBufferCount(params_.src_bufs, 0);
-    assertBufferCount(params_.dst_bufs, 1);
-  }
-
-  if (params_.team.size() == 1) {
-    return nullptr;
-  }
-
-  return comm.getBackendForTeam(params_.team, backend)
-      ->broadcast(
-          comm.deviceId() == params_.root ? params_.src_bufs : params_.dst_bufs,
-          {.rootRank = root_relative_index_});
-}
-
-Gather::Gather(CommParams params) : Communication(params, "gather") {
-  assertBufferCount(params_.src_bufs, 1);
+std::string Communication::toInlineString(int indent_size) const {
+  return toString(indent_size);
 }
 
-c10::intrusive_ptr<c10d::Work> Gather::post(
-    Communicator& comm,
-    std::optional<CommunicatorBackend> backend) {
-  post_common(*this, comm);
-  // This is used to change the representation of the buffers to match c10d
-  // ProcessGroup API
-  std::vector<std::vector<at::Tensor>> buf_list;
-  if (comm.deviceId() == params_.root) {
-    assertBufferCount(params_.dst_bufs, params_.team.size());
-    buf_list = {std::move(params_.dst_bufs)};
-  } else {
-    assertBufferCount(params_.dst_bufs, 0);
+// TODO add checking symbolic representation of src and dst buffers
+bool Communication::sameAs(const Statement* other) const {
```

Comment:
I failed to see why we couldn't reuse the "default" implementation: https://github.com/NVIDIA/Fuser/blob/4e8356915fc6ffa823cf97fed4249cc00f317fdf/csrc/ir/base_nodes.cpp#L342. Can you clarify?

Reply:
Two Communications could be "Expr::sameAs" without being the same, for example if one is Allgather and the other Allreduce. Right?
My goal here is to mimick what's done for other IRs, such as https://github.com/NVIDIA/Fuser/blob/4e8356915fc6ffa823cf97fed4249cc00f317fdf/csrc/ir/nodes.cpp#L2674

But I don't actually use sameAs -- I simply thought implementing it was necessary for matching the IR specs.

So I'm open to any suggestion about this implementation, including removing it

Reply:
`Expr::sameAs` calls `Expr::sameOp`, which checks equality of all attributes. I believe the communication type is one of the attributes.

Reply:
Correct me if Im wrong, but in the current implementation it is not an attribute. The reason is that, IIUC, all attributes need to be `Statement*`s, and `CommunicationParams` is not.

Reply:
There is addDataAttribute as used in https://github.com/NVIDIA/Fuser/blob/730abb552e5e95ac3476a2a6e9b60bfe114de2c0/csrc/ir/nodes.cpp#L2423-L2424. But the PR as is is already hard to merge, I'll clean that up in a separate PR.